ledger/participant-integration-api/src/main/scala/platform/DispatcherState.scala

```diff
@@ -24,53 +24,54 @@ class DispatcherState(dispatcherShutdownTimeout: Duration)(implicit
+  def getDispatcher: Dispatcher[Offset] = dispatcherStateRef match {
+    case DispatcherNotRunning => throw dispatcherNotRunning()
+    case DispatcherStateShutdown => throw dispatcherStateShutdown()
+    case DispatcherRunning(dispatcher) => dispatcher
   }
 
-  def getDispatcher: Dispatcher[Offset] = dispatcherStateRef match {
-    case DispatcherState.DispatcherNotInitialized => throw dispatcherNotInitializedException()
-    case DispatcherState.DispatcherOperational(dispatcher) => dispatcher
-    case DispatcherState.DispatcherShutdown => throw dispatcherShutdownException()
+  def startDispatcher(ledgerEnd: LedgerEnd): Unit = synchronized {
+    dispatcherStateRef match {
+      case DispatcherNotRunning =>
+        val activeDispatcher = buildDispatcher(ledgerEnd)
+        dispatcherStateRef = DispatcherRunning(activeDispatcher)
+      case DispatcherStateShutdown => throw dispatcherStateShutdown()
+      case DispatcherRunning(_) =>
+        throw new IllegalStateException(
+          "Dispatcher startup triggered while an existing dispatcher is still active."
+        )
+    }
   }
 
-  def reset(ledgerEnd: LedgerEnd): Future[Unit] = synchronized {
+  def stopDispatcher(): Future[Unit] = synchronized {
     dispatcherStateRef match {
-      case DispatcherState.DispatcherNotInitialized =>
-        dispatcherStateRef = DispatcherOperational(buildDispatcher(ledgerEnd))
-        Future.unit
-      case DispatcherOperational(dispatcher) =>
-        dispatcherStateRef = DispatcherOperational(buildDispatcher(ledgerEnd))
-        shutdownDispatcher(dispatcher)
-      case DispatcherState.DispatcherShutdown =>
-        Future.failed(dispatcherShutdownException())
+      case DispatcherNotRunning | DispatcherStateShutdown => Future.unit
+      case DispatcherRunning(dispatcher) =>
+        dispatcherStateRef = DispatcherNotRunning
+        logger.info("Stopping active Ledger API offset dispatcher.")
+        dispatcher
+          // TODO LLP: Fail sources with exception instead of graceful shutdown
+          .shutdown()
+          .withTimeout(dispatcherShutdownTimeout)(
+            logger.warn(
+              s"Shutdown of existing Ledger API streams did not finish in ${dispatcherShutdownTimeout.toSeconds} seconds."
+            )
+          )
     }
   }
 
   private[platform] def shutdown(): Future[Unit] = synchronized {
-    val currentState = dispatcherStateRef
-    dispatcherStateRef = DispatcherShutdown
-
-    currentState match {
-      case DispatcherState.DispatcherNotInitialized | DispatcherState.DispatcherShutdown =>
-        Future.unit
-      case DispatcherOperational(dispatcher) =>
-        shutdownDispatcher(dispatcher)
+    logger.info("Shutting down Ledger API offset dispatcher state.")
+    val shutdownF = dispatcherStateRef match {
+      case DispatcherNotRunning | DispatcherStateShutdown => Future.unit
+      case DispatcherRunning(_) => stopDispatcher()
```

Comment:
Calling the public method repeats some checks that have just been made.  Maybe a nicer way would be to have a low level `stopRunningDispatcher` that is called from both `shutdown` and `stopDispatcher` ?

Reply:
Refactored so there is no need for a common method anymore